ethcore/src/trace/db.rs

```diff
@@ -0,0 +1,573 @@
+// Copyright 2015, 2016 Ethcore (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Fat database.
+use std::collections::HashMap;
+use std::sync::{RwLock, Arc};
+use std::path::Path;
+use bloomchain::Number;
+use bloomchain::group::{BloomGroupDatabase, BloomGroupChain, GroupPosition, BloomGroup};
+use util::{FixedHash, H256, H264, Database, DBTransaction};
+use header::BlockNumber;
+use trace::{BlockTraces, LocalizedTrace, Config, Filter, Database as TraceDatabase, ImportRequest,
+DatabaseExtras};
+use db::{Key, Writable, Readable, CacheUpdatePolicy};
+use super::bloom::{TraceGroupPosition, BlockTracesBloom, BlockTracesBloomGroup};
+use super::flat::{FlatTrace, FlatBlockTraces, FlatTransactionTraces};
+
+#[derive(Debug, Copy, Clone)]
+pub enum TracedbIndex {
+	/// Block traces index.
+	BlockTraces = 0,
+	/// Trace bloom group index.
+	BlockTracesBloomGroups = 1,
+}
+
+fn with_index(hash: &H256, i: TracedbIndex) -> H264 {
+	let mut slice = H264::from_slice(hash);
+	slice[32] = i as u8;
+	slice
+}
+
+impl Key<BlockTraces> for H256 {
+	fn key(&self) -> H264 {
+		with_index(self, TracedbIndex::BlockTraces)
+	}
+}
+
+impl Key<BlockTracesBloomGroup> for TraceGroupPosition {
+	fn key(&self) -> H264 {
+		with_index(&self.hash(), TracedbIndex::BlockTracesBloomGroups)
+	}
+}
+
+/// Fat database.
+pub struct Tracedb<T> where T: DatabaseExtras {
+	// cache
+	traces: RwLock<HashMap<H256, BlockTraces>>,
+	blooms: RwLock<HashMap<TraceGroupPosition, BlockTracesBloomGroup>>,
+	// db
+	tracesdb: Database,
+	// config,
+	config: Config,
+	// extras
+	extras: Arc<T>,
+}
+
+impl<T> BloomGroupDatabase for Tracedb<T> where T: DatabaseExtras {
+	fn blooms_at(&self, position: &GroupPosition) -> Option<BloomGroup> {
+		let position = TraceGroupPosition::from(position.clone());
+		self.tracesdb.read_with_cache(&self.blooms, &position).map(Into::into)
+	}
+}
+
+impl<T> Tracedb<T> where T: DatabaseExtras {
+	/// Creates new instance of `Tracedb`.
+	pub fn new(mut config: Config, path: &Path, extras: Arc<T>) -> Self {
+		let mut tracedb_path = path.to_path_buf();
+		tracedb_path.push("tracedb");
+		let tracesdb = Database::open_default(tracedb_path.to_str().unwrap()).unwrap();
```

Comment:
what's the idea behind having a separate DB? are we going to start having loads of different DBs now? VM tracing, and fat db (i.e. hash preimages) are both independent additions. is there a storage/memory/compute/file-handle cost to loads of open databases?